### ISSUE TITLE
cisco_nxos show ip interface - TextFSMError

### ISSUE THREAD

**Issue by georgealmg** [networktocode/ntc-templates#1027]
##### ISSUE TYPE

 - Feature Idea

##### TEMPLATE USING

Value Filldown VRF_NAME (\S+)
Value Required INTERFACE (\S+)
Value PROTOCOL_STATUS (\w+)
Value LINK_STATUS (\w+)
Value ADMIN_STATUS (\w+)
Value PRIMARY_IP_ADDRESS (\S+)
Value PRIMARY_IP_SUBNET (\S+)
Value PRIMARY_BROADCAST_ADDRESS (\S+)
Value List SECONDARY_IP_ADDRESS (\S+)
Value List SECONDARY_IP_SUBNET (\S+)
Value MULTICAST_GROUPS (224\..*?)
Value MTU (\d+)
Value ROUTE_PREFERENCE (\d+)
Value ROUTE_PREFERENCE_TAG (\d+)
Value PROXY_ARP (\S+)
Value LOCAL_PROXY_ARP (\S+)
Value MULTICAST_ROUTING (\S+)
Value ICMP_REDIRECTS (\S+)
Value DIRECTED_BROADCAST (\S+)
Value IP_FORWARDING (\S+)
Value ICMP_UNREACHABLES (\S+)
Value ICMP_PORT_UNREACHABLE (\S+)
Value URPF (\S+)
Value LOAD_SHARING (\S+)
Value LAST_RESET (\S+)
Value UNICAST_PACKETS_SENT (\d+)
Value UNICAST_PACKETS_RECEIVED (\d+)
Value UNICAST_PACKETS_FORWARDED (\d+)
Value UNICAST_PACKETS_ORIGINATED (\d+)
Value UNICAST_PACKETS_CONSUMED (\d+)
Value UNICAST_BYTES_SENT (\d+)
Value UNICAST_BYTES_RECEIVED (\d+)
Value UNICAST_BYTES_FORWARDED (\d+)
Value UNICAST_BYTES_ORIGINATED (\d+)
Value UNICAST_BYTES_CONSUMED (\d+)
Value MULTICAST_PACKETS_SENT (\d+)
Value MULTICAST_PACKETS_RECEIVED (\d+)
Value MULTICAST_PACKETS_FORWARDED (\d+)
Value MULTICAST_PACKETS_ORIGINATED (\d+)
Value MULTICAST_PACKETS_CONSUMED (\d+)
Value MULTICAST_BYTES_SENT (\d+)
Value MULTICAST_BYTES_RECEIVED (\d+)
Value MULTICAST_BYTES_FORWARDED (\d+)
Value MULTICAST_BYTES_ORIGINATED (\d+)
Value MULTICAST_BYTES_CONSUMED (\d+)
Value BROADCAST_PACKETS_SENT (\d+)
Value BROADCAST_PACKETS_RECEIVED (\d+)
Value BROADCAST_PACKETS_FORWARDED (\d+)
Value BROADCAST_PACKETS_ORIGINATED (\d+)
Value BROADCAST_PACKETS_CONSUMED (\d+)
Value BROADCAST_BYTES_SENT (\d+)
Value BROADCAST_BYTES_RECEIVED (\d+)
Value BROADCAST_BYTES_FORWARDED (\d+)
Value BROADCAST_BYTES_ORIGINATED (\d+)
Value BROADCAST_BYTES_CONSUMED (\d+)
Value LABELED_PACKETS_SENT (\d+)
Value LABELED_PACKETS_RECEIVED (\d+)
Value LABELED_PACKETS_FORWARDED (\d+)
Value LABELED_PACKETS_ORIGINATED (\d+)
Value LABELED_PACKETS_CONSUMED (\d+)
Value LABELED_BYTES_SENT (\d+)
Value LABELED_BYTES_RECEIVED (\d+)
Value LABELED_BYTES_FORWARDED (\d+)
Value LABELED_BYTES_ORIGINATED (\d+)
Value LABELED_BYTES_CONSUMED (\d+)
Value WCCP_REDIRECT_OUTBOUND (\S+)
Value WCCP_REDIRECT_INBOUND (\S+)
Value WCCP_REDIRECT_EXCLUDE (\S+)

Start
  ^IP\s+Interface\s+Status\s+for\s+VRF\s+\"${VRF_NAME}\"\s*$$
  ^${INTERFACE},.*protocol-${PROTOCOL_STATUS}\/link-${LINK_STATUS}\/admin-${ADMIN_STATUS},.*$$
  ^\s+IP\s+address:\s+${PRIMARY_IP_ADDRESS},\s+IP\s+subnet:\s+${PRIMARY_IP_SUBNET}\s*$$
  # Secondary IP addresses are stored as a list
  ^\s+IP\s+address:\s+${SECONDARY_IP_ADDRESS},\s+IP\s+subnet:\s+${SECONDARY_IP_SUBNET}\s+secondary\s*$$
  ^\s+IP\s+broadcast\s+address:\s+${PRIMARY_BROADCAST_ADDRESS}\s*$$
  ^\s+IP\s+multicast\s+groups\s+locally\s+joined:.*$$
  # Multiple multicast groups are on the same line so we can't make a list out of those
  ^\s+${MULTICAST_GROUPS}\s*$$
  ^\s+IP\s+MTU:\s+${MTU}.*$$
  ^\s+IP\s+primary\s+address\s+route-preference:\s+${ROUTE_PREFERENCE}, tag:\s+${ROUTE_PREFERENCE_TAG}\s*$$
  ^\s+IP\s+proxy\s+ARP\s*:\s+${PROXY_ARP}\s*$$
  ^\s+IP\s+Local\s+Proxy\s+ARP\s*:\s+${LOCAL_PROXY_ARP}\s*$$
  ^\s+IP\s+multicast\s+routing:\s+${MULTICAST_ROUTING}\s*$$
  ^\s+IP\s+icmp\s+redirects:\s+${ICMP_REDIRECTS}\s*$$
  ^\s+IP\s+directed-broadcast:\s+${DIRECTED_BROADCAST}(,.*|\s*)$$
  ^\s+IP\s+Forwarding:\s+${IP_FORWARDING}(,.*|\s*)$$
  ^\s+IP\s+icmp\s+unreachables \(except port\):\s+${ICMP_UNREACHABLES}\s*$$
  ^\s+IP\s+icmp\s+port-unreachable:\s+${ICMP_PORT_UNREACHABLE}\s*$$
  ^\s+IP\s+unicast\s+reverse\s+path\s+forwarding:\s+${URPF}\s*$$
  ^\s+IP\s+load\s+sharing:\s+${LOAD_SHARING}\s*$$
  ^\s+IP\s+interface\s+statistics\s+last\s+reset:\s+${LAST_RESET}\s*$$
  ^\s+IP\s+interface\s+software\s+stats:.*$$
  ^\s+Unicast\s+packets\s+:\s+${UNICAST_PACKETS_SENT}\/${UNICAST_PACKETS_RECEIVED}\/${UNICAST_PACKETS_FORWARDED}\/${UNICAST_PACKETS_ORIGINATED}\/${UNICAST_PACKETS_CONSUMED}\s*$$
  ^\s+Unicast\s+bytes\s+:\s+${UNICAST_BYTES_SENT}\/${UNICAST_BYTES_RECEIVED}\/${UNICAST_BYTES_FORWARDED}\/${UNICAST_BYTES_ORIGINATED}\/${UNICAST_BYTES_CONSUMED}\s*$$
  ^\s+Multicast\s+packets\s+:\s+${MULTICAST_PACKETS_SENT}\/${MULTICAST_PACKETS_RECEIVED}\/${MULTICAST_PACKETS_FORWARDED}\/${MULTICAST_PACKETS_ORIGINATED}\/${MULTICAST_PACKETS_CONSUMED}\s*$$
  ^\s+Multicast\s+bytes\s+:\s+${MULTICAST_BYTES_SENT}\/${MULTICAST_BYTES_RECEIVED}\/${MULTICAST_BYTES_FORWARDED}\/${MULTICAST_BYTES_ORIGINATED}\/${MULTICAST_BYTES_CONSUMED}\s*$$
  ^\s+Broadcast\s+packets\s+:\s+${BROADCAST_PACKETS_SENT}\/${BROADCAST_PACKETS_RECEIVED}\/${BROADCAST_PACKETS_FORWARDED}\/${BROADCAST_PACKETS_ORIGINATED}\/${BROADCAST_PACKETS_CONSUMED}\s*$$
  ^\s+Broadcast\s+bytes\s+:\s+${BROADCAST_BYTES_SENT}\/${BROADCAST_BYTES_RECEIVED}\/${BROADCAST_BYTES_FORWARDED}\/${BROADCAST_BYTES_ORIGINATED}\/${BROADCAST_BYTES_CONSUMED}\s*$$
  ^\s+Labeled\s+packets\s+:\s+${LABELED_PACKETS_SENT}\/${LABELED_PACKETS_RECEIVED}\/${LABELED_PACKETS_FORWARDED}\/${LABELED_PACKETS_ORIGINATED}\/${LABELED_PACKETS_CONSUMED}\s*$$
  ^\s+Labeled\s+bytes\s+:\s+${LABELED_BYTES_SENT}\/${LABELED_BYTES_RECEIVED}\/${LABELED_BYTES_FORWARDED}\/${LABELED_BYTES_ORIGINATED}\/${LABELED_BYTES_CONSUMED}\s*$$
  ^\s+WCCP\s+Redirect\s+outbound:\s+${WCCP_REDIRECT_OUTBOUND}\s*$$
  ^\s+WCCP\s+Redirect\s+inbound:\s+${WCCP_REDIRECT_INBOUND}\s*$$
  ^\s+WCCP\s+Redirect\s+exclude:\s+${WCCP_REDIRECT_EXCLUDE}\s*$$ -> Record
  ^\s*$$
  ^. -> Error

##### SAMPLE COMMAND OUTPUT

show ip interface
IP Interface Status for VRF "default"
Vlanxx, Interface status: protocol-up/link-up/admin-up, iod: 126,
  IP address: x.x.x.x, IP subnet: x.x.x.x/xx route-preference: 0, tag: 0
  IP broadcast address: 255.255.255.255
  IP multicast groups locally joined: 
      224.0.0.2  
  IP MTU: 1500 bytes (using link MTU)
  IP primary address route-preference: 0, tag: 0
  IP proxy ARP : disabled
  IP Local Proxy ARP : disabled
  IP multicast routing: disabled
  IP icmp redirects: disabled
  IP directed-broadcast: disabled 
  IP Forwarding: disabled 
  IP icmp unreachables (except port): disabled
  IP icmp port-unreachable: enabled
  IP unicast reverse path forwarding: none
  IP load sharing: none 
  IP interface statistics last reset: never
  IP interface software stats: (sent/received/forwarded/originated/consumed)
    Unicast packets    : 3709/14997382/0/3739/29995645
    Unicast bytes      : 430030/1260055911/0/432550/2520162214
    Multicast packets  : 0/63088860/0/0/126177720
    Multicast bytes    : 0/3028265280/0/0/3028265280
    Broadcast packets  : 0/0/0/0/0
    Broadcast bytes    : 0/0/0/0/0
    Labeled packets    : 0/0/0/0/0
    Labeled bytes      : 0/0/0/0/0
  WCCP Redirect outbound: disabled
  WCCP Redirect inbound: disabled
  WCCP Redirect exclude: disabled

##### SUMMARY

While trying to use the show ip interface template on a nexus 9000 I got a TextFSM Error referencing the following line:

 "IP address: x.x.x.x, IP subnet: x.x.x.x/xx route-preference: 0, tag: 0"

Looking at line 73 of the template I can see that the issue is that the template isn't compatible so maybe support for this new line format could be added.

##### STEPS TO REPRODUCE

On a nexus 9000:
parse_output(platform="cisco_nxos",command="show ip interface",data=conn.send_command("show ip interface"))

##### EXPECTED RESULTS

Parsed output.

##### ACTUAL RESULTS

TextFSMError.

**Comment by diepes** [comment]
create PR https://github.com/networktocode/ntc-templates/pull/1046

**Comment by Mattias3891** [comment]
I hit this bug too. Puzzled.
In the template, adding .* regex to Prim IP subnet line seems to solve it. 
`
 ^\s+IP\s+address:\s+${PRIMARY_IP_ADDRESS},\s+IP\s+subnet:\s+${PRIMARY_IP_SUBNET}\s*$$
`
to
` 
^\s+IP\s+address:\s+${PRIMARY_IP_ADDRESS},\s+IP\s+subnet:\s+${PRIMARY_IP_SUBNET}\s*.*$$
`

**Comment by Mattias3891** [comment]
> 

Well. No wonder I ended up with issues. n00b error. Overwrote my primary IPs with secondary IPs.

This seems to be better. It tests against outputs with route pref and without. 
```
^\s+IP\s+address:\s+${PRIMARY_IP_ADDRESS},\s+IP\s+subnet:\s+${PRIMARY_IP_SUBNET}\s*$$
  ^\s+IP\s+address:\s+${PRIMARY_IP_ADDRESS},\s+IP\s+subnet:\s+${PRIMARY_IP_SUBNET}\s+route.*$$
  # Secondary IP addresses are stored as a list
  ^\s+IP\s+address:\s+${SECONDARY_IP_ADDRESS},\s+IP\s+subnet:\s+${SECONDARY_IP_SUBNET}\s+secondary\s*$$
  ^\s+IP\s+address:\s+${SECONDARY_IP_ADDRESS},\s+IP\s+subnet:\s+${SECONDARY_IP_SUBNET}\s+secondary\s+route.*$$
```

**Comment by ncutting** [comment]
Mattias3891's last post fixed the error.  However I only ever get the last interface returned in the list?

This is also true for cisco_nxos_show_ip_interface_vrf_all.textfsm - How to get the list of all interfaces? are they being overwritten each time?

**Comment by itdependsnetworks** [comment]
Fixed in #1046
